File: modules/project/src/main/java/com/falsepattern/zigbrains/project/toolchain/LocalZigToolchain.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.project.toolchain;

import com.falsepattern.zigbrains.common.direnv.DirenvCmd;
import com.falsepattern.zigbrains.common.util.PathUtil;
import com.falsepattern.zigbrains.project.openapi.components.ZigProjectSettingsService;
import com.intellij.execution.ExecutionException;
import com.intellij.execution.configurations.GeneralCommandLine;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.UserDataHolder;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.nio.file.Path;

public class LocalZigToolchain extends AbstractZigToolchain{
    public LocalZigToolchain(Path location, @Nullable Project project) {
        super(location, project);
    }

    @Override
    public int executionTimeoutInMilliseconds() {
        return 10000;
    }

    @Override
    public @NotNull GeneralCommandLine patchCommandLine(@NotNull GeneralCommandLine commandLine, @NotNull UserDataHolder data) {
        val project = getProject();
        val direnv = data.getUserData(DirenvCmd.DIRENV_KEY);
        if (direnv != null && direnv) {
            commandLine.withEnvironment(DirenvCmd.tryGetProjectEnvSync(project));
        }
        return commandLine;
    }

    @Override
    public Path pathToExecutable(String toolName) {
        return PathUtil.pathToExecutable(getLocation(), toolName);
    }

    @Override
    public @NotNull UserDataHolder getDataForSelfRuns() {
        val data = super.getDataForSelfRuns();
        val project = getProject();
        if (project != null) {
            data.putUserData(DirenvCmd.DIRENV_KEY, ZigProjectSettingsService.getInstance(project).getState().direnv);
        }
        return data;
    }

    public static LocalZigToolchain ensureLocal(AbstractZigToolchain toolchain) throws ExecutionException {
        if (!(toolchain instanceof LocalZigToolchain $toolchain)) {
            throw new ExecutionException("The debugger only supports local zig toolchains!");
        }
        return $toolchain;
    }
}

File: core/src/main/kotlin/com/falsepattern/zigbrains/project/toolchain/LocalZigToolchain.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.project.toolchain

import com.falsepattern.zigbrains.direnv.DirenvCmd
import com.falsepattern.zigbrains.project.settings.zigProjectSettings
import com.intellij.execution.ExecutionException
import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.guessProjectDir
import com.intellij.openapi.util.KeyWithDefaultValue
import com.intellij.openapi.util.SystemInfo
import com.intellij.openapi.util.io.toNioPathOrNull
import com.intellij.openapi.vfs.toNioPathOrNull
import java.nio.file.Path
import java.util.UUID
import kotlin.io.path.pathString

data class LocalZigToolchain(val location: Path, val std: Path? = null, val name: String? = null): AbstractZigToolchain() {
    override fun workingDirectory(project: Project?): Path? {
        return project?.guessProjectDir()?.toNioPathOrNull()
    }

    override suspend fun patchCommandLine(commandLine: GeneralCommandLine, project: Project?): GeneralCommandLine {
        if (project != null && (commandLine.getUserData(DIRENV_KEY) ?: project.zigProjectSettings.state.direnv)) {
            commandLine.withEnvironment(DirenvCmd.importDirenv(project).env)
        }
        return commandLine
    }

    override fun pathToExecutable(toolName: String, project: Project?): Path {
        val exeName = if (SystemInfo.isWindows) "$toolName.exe" else toolName
        return location.resolve(exeName)
    }

    companion object {
        val DIRENV_KEY = KeyWithDefaultValue.create<Boolean>("ZIG_LOCAL_DIRENV")

        @Throws(ExecutionException::class)
        fun ensureLocal(toolchain: AbstractZigToolchain): LocalZigToolchain {
            if (toolchain is LocalZigToolchain) {
                return toolchain
            } else {
                // TODO
                throw ExecutionException("The debugger only supports local zig toolchain")
            }
        }

        fun tryFromPathString(pathStr: String): LocalZigToolchain? {
            return pathStr.toNioPathOrNull()?.let(::tryFromPath)
        }

        fun tryFromPath(path: Path): LocalZigToolchain? {
            val tc = LocalZigToolchain(path)
            if (!tc.zig.fileValid()) {
                return null
            }
            return tc
        }
    }
}